pkg/redis/client.go

```diff
@@ -26,43 +29,53 @@ type Client struct {
 	logger *logging.Logger
 }
 
-// Options define user configurable Redis options.
-type Options struct {
-	BlockTimeout        time.Duration `yaml:"block_timeout"         default:"1s"`
-	HMGetCount          int           `yaml:"hmget_count"           default:"4096"`
-	HScanCount          int           `yaml:"hscan_count"           default:"4096"`
-	MaxHMGetConnections int           `yaml:"max_hmget_connections" default:"8"`
-	Timeout             time.Duration `yaml:"timeout"               default:"30s"`
-	XReadCount          int           `yaml:"xread_count"           default:"4096"`
+// NewClient returns a new Client wrapper for a pre-existing redis.Client.
+func NewClient(client *redis.Client, logger *logging.Logger, options *Options) *Client {
+	return &Client{Client: client, logger: logger, Options: options}
 }
 
-// Validate checks constraints in the supplied Redis options and returns an error if they are violated.
-func (o *Options) Validate() error {
-	if o.BlockTimeout <= 0 {
-		return errors.New("block_timeout must be positive")
-	}
-	if o.HMGetCount < 1 {
-		return errors.New("hmget_count must be at least 1")
-	}
-	if o.HScanCount < 1 {
-		return errors.New("hscan_count must be at least 1")
+// NewClientFromConfig returns a new Client from Config.
+func NewClientFromConfig(c *Config, logger *logging.Logger) (*Client, error) {
+	tlsConfig, err := c.TlsOptions.MakeConfig(c.Host)
+	if err != nil {
+		return nil, err
 	}
-	if o.MaxHMGetConnections < 1 {
-		return errors.New("max_hmget_connections must be at least 1")
+
+	var dialer ctxDialerFunc
+	dl := &net.Dialer{Timeout: 15 * time.Second}
+
+	if tlsConfig == nil {
+		dialer = dl.DialContext
+	} else {
+		dialer = (&tls.Dialer{NetDialer: dl, Config: tlsConfig}).DialContext
 	}
-	if o.Timeout == 0 {
-		return errors.New("timeout cannot be 0. Configure a value greater than zero, or use -1 for no timeout")
+
+	options := &redis.Options{
+		Dialer:      dialWithLogging(dialer, logger),
+		Password:    c.Password,
+		DB:          0, // Use default DB,
+		ReadTimeout: c.Options.Timeout,
+		TLSConfig:   tlsConfig,
 	}
-	if o.XReadCount < 1 {
-		return errors.New("xread_count must be at least 1")
+
+	if utils.IsUnixAddr(c.Host) {
+		options.Network = "unix"
+		options.Addr = c.Host
+	} else {
+		port := c.Port
+		if port == 0 {
+			port = 6379
+		}
```

Comment:
The default value does not match the one set in `config#SetDefaults()`, and that variable is never used afterwards. However, do you even need this check at all, given that `config#SetDefaults()` takes care of it? It might also be reasonable to let Icinga DB fail in `Redis#Config#Validate()` if the user explicitly set this to `0`, as this is not a valid port.